src/Stomp/Network/Observer/Heartbeat/Emitter.php

```diff
@@ -0,0 +1,270 @@
+    }
+
+    /**
+     * Indicates that a frame has been received from the server.
+     *
+     * @param Frame $frame that has been received
+     * @return void
+     */
+    public function receivedFrame(Frame $frame)
+    {
+        if ($frame->getCommand() === Emitter::FRAME_SERVER_CONNECTED) {
+            $beats = $this->getHeartbeats($frame);
+            $this->intervalServer = $beats[1];
+            if ($this->intervalServer && ($this->intervalClient || $this->intervalClient === null)) {
+                $this->interval = $this->intervalUsage * (max($this->intervalServer,
+                            $this->intervalClient) / 1000); // milli to micro
+                $this->enabled = true;
+            }
+        } else {
+            $this->onPassiveEvent();
+        }
+    }
+
+    /**
+     * Returns the heart beat header.
+     *
+     * @param Frame $frame
+     * @return array
+     */
+    private function getHeartbeats(Frame $frame)
+    {
+        $beats = $frame['heart-beat'];
+        if ($beats) {
+            return explode(',', $beats, 2);
+        }
+        return [0, 0];
+    }
+
+    /**
+     * Indicates that a frame has been sent to the server.
+     *
+     * @param Frame $frame
+     * @return void
+     */
+    public function sentFrame(Frame $frame)
+    {
+        if ($this->enabled) {
+            $this->notifyBeat();
+            return;
+        }
+        if ($frame->getCommand() === Emitter::FRAME_CLIENT_CONNECT) {
+            $beats = $this->getHeartbeats($frame);
+            $this->intervalClient = $beats[0];
+            $this->notifyBeat();
+        }
+    }
+
+    /**
+     * Returns the microtime for the moment when the last outgoing beat was detected.
+     *
+     * @return float
+     */
+    public function getLastbeat()
+    {
+        return $this->lastbeat;
+    }
+
+    /**
+     * Returns the calculated interval for outgoing beats in seconds (with micro fraction).
+     *
+     * @return float
+     */
+    public function getInterval()
+    {
+        return $this->interval;
+    }
+
+    /**
+     * Checks if the emitter was enabled.
+     *
+     * @return bool
+     */
+    public function isEnabled()
+    {
+        return $this->enabled;
+    }
+
+    /**
+     * Returns the interval usage.
+     *
+     * @return float
+     */
+    public function getIntervalUsage()
+    {
+        return $this->intervalUsage;
+    }
+
+    /**
+     * Sets the interval usage, must be configured before the emitter gets active to work.
+     *
+     * The usage (percentage) defines the amount of agreed beat interval time,
+     * that is allowed to pass before the emitter will send a beat.
+     *
+     * A higher value increases the risk that a beat is send after a timeout has occurred.
+     * A lower value increases the beats and adds overhead to the connection.
+     *
+     * @param float $intervalUsage
+     */
+    public function setIntervalUsage($intervalUsage)
```

Comment:
as this value should be set before the emitter gets active, should it be a constructor arg instead?